tensorstore/kvstore/s3/aws_metadata_credential_provider.cc

```diff
@@ -0,0 +1,202 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include <optional>
+#include <string>
+
+#include "absl/time/time.h"
+#include "absl/strings/str_cat.h"
+#include "tensorstore/util/status.h"
+#include "tensorstore/util/result.h"
+#include "tensorstore/internal/http/http_request.h"
+#include "tensorstore/internal/http/http_response.h"
+#include "tensorstore/internal/http/http_transport.h"
+#include "tensorstore/kvstore/s3/aws_metadata_credential_provider.h"
+#include "tensorstore/internal/json_binding/absl_time.h"
+#include "tensorstore/internal/json_binding/bindable.h"
+#include "tensorstore/internal/json_binding/std_optional.h"
+#include "tensorstore/internal/json_binding/json_binding.h"
+#include "tensorstore/util/str_cat.h"
+
+using ::tensorstore::Result;
+using ::tensorstore::internal::ParseJson;
+using ::tensorstore::internal_http::HttpRequestBuilder;
+using ::tensorstore::internal_http::HttpResponseCodeToStatus;
+using ::tensorstore::internal_kvstore_s3::AwsCredentials;
+
+namespace jb = tensorstore::internal_json_binding;
+
+namespace tensorstore {
+namespace internal_kvstore_s3 {
+
+namespace {
+
+// Token ttl header
+static constexpr char kTokenTtlHeader[] = "x-aws-ec2-metadata-token-ttl-seconds";
+// Token header
+static constexpr char kMetadataTokenHeader[] = "x-aws-ec2-metadata-token";
+// Obtain Metadata server API tokens from this url
+static constexpr char kTokenUrl[] = "http://http://169.254.169.254/latest/api/token";
+// Obtain IAM status from this url
+static constexpr char kIamUrl[] = "http://http://169.254.169.254/latest/meta-data/iam/";
+// Obtain current IAM role from this url
+static constexpr char kIamCredentialsUrl[] = "http://169.254.169.254/latest/meta-data/iam/security-credentials/";
+
+// Requests to the above server block outside AWS
+// Configure a timeout small enough not to degrade performance outside AWS
+// but large enough to give the EC2Metadata enough time to respond
+static constexpr absl::Duration kConnectTimeout = absl::Milliseconds(200);
+
+// Successful EC2Metadata Security Credential Response Code
+static constexpr char kSuccess[] = "Success";
+
+/// Represents JSON returned from
+/// http://http://169.254.169.254/latest/meta-data/iam/security-credentials/<iam-role>/
+/// where <iam-role> is usually returned as a response from a request to
+/// http://http://169.254.169.254/latest/meta-data/iam/security-credentials/
+struct EC2CredentialsResponse {
+  std::string Code;
+  std::optional<absl::Time> LastUpdated;
+  std::optional<std::string> Type;
+  std::optional<std::string> AccessKeyId;
+  std::optional<std::string> SecretAccessKey;
+  std::optional<std::string> Token;
+  std::optional<absl::Time> Expiration;
+
+  using ToJsonOptions = IncludeDefaults;
```

Comment:
Since this isn't exposed to other code, you don't need all the default binder stuff (lines 78-84, lines 97-102).

You already have

```
inline constexpr auto EC2CredentialsResponseBinder = ...
```

So instead of ` EC2CredentialsResponse::FromJson(ParseJson(json_sv))`

You can use this: `jb::FromJson<EC2CredentialsResponse>(json_sv, EC2CredentialsResponseBinder)`

Reply:
Thanks, used this suggestion.